### PR TITLE
[Model Monitoring]: Create model monitoring feature set per model endpoint

### PR DESCRIPTION
**Pull request by EyalDanieliIgu**
Instead of using general model monitoring feature per **project**, we create model monitoring feature set per **model endpoint**. The monitoring graph that was defined in the general monitoring feature set is now converted into a serving graph within the model monitoring stream.
	
**With this new logic:**
- The user can decide for each model endpoint whether to apply model monitoring or not.
- Each model endpoint has a new flag called 'monitoring' (under `model_endpoint.spec.monitoring`). If `model_endpoint.spec.monitoring` is True then once we deploy the model endpoint, we also deploy the model monitoring feature set. The name and the type of each feature is taken from the model object. Note that if it's not exist within the model object, the monitoring feature set will be empty of features. 
- The parquet target files that are generated from the model monitoring stream are now partitioned by model endpoint. In other words, each parquet file is connected to a single model endpoint and includes the relevant schema. It allows the model monitoring batch job to process these parquet files by model endpoint and to avoid unnecessary repetitive analysis.
- After each batch process (hourly job by default), the new model monitoring feature set provides an updated statistics based on the latest batch. 
	 
	
In addition, the following PR also includes a fix for handling the error of deploying model monitoring batch job after it has been already deployed.

### CODE REVIEW

File: mlrun/api/crud/model_endpoints.py [hunk 1]
```diff
@@ -120,7 +137,83 @@ def create_or_patch(
...
+        Create monitoring feature set with the relevant parquet target.
+
+        :param model_endpoint:    An object representing the model endpoint
+        :param model_obj:         An object representing the deployed model
+        :param db_session:        A session that manages the current dialog with the database
+
+        """
+
+        # define a new feature set
+        serving_function_name = model_endpoint.spec.function_uri.replace(
```
Comment (yaronha) [inline mlrun/api/crud/model_endpoints.py:156]: use `parse_versioned_object_uri()` instead

File: mlrun/api/crud/model_endpoints.py [hunk 2]
```diff
@@ -120,7 +137,83 @@ def create_or_patch(
...
+        )
+        model_name = model_endpoint.spec.model.replace(":", "-")
+
+        feature_set = mlrun.feature_store.FeatureSet(
+            f"monitoring-{serving_function_name}-{model_name}",
+            entities=["endpoint_id"],
+            timestamp_key="timestamp",
+            description=f"Monitoring feature set for endpoint: {model_endpoint.spec.model}",
+        )
+        feature_set.metadata.project = model_endpoint.metadata.project
```
Comment (yaronha) [inline mlrun/api/crud/model_endpoints.py:167]: i suggest to add `labels` w relevant info e.g. `feature_set.metadata.label["endpoint"] = endpoint_id`

File: mlrun/api/crud/model_endpoints.py [hunk 3]
```diff
@@ -120,7 +137,83 @@ def create_or_patch(
...
+
+        # add features to the feature set according to the model object
+        if len(model_obj.inputs.to_dict()) > 0:
```
Comment (yaronha) [inline mlrun/api/crud/model_endpoints.py:170]: can simply `len(model_obj.inputs)` or `model_obj.inputs.values()`

File: mlrun/api/crud/model_endpoints.py [hunk 4]
```diff
@@ -120,7 +137,83 @@ def create_or_patch(
...
+                        name=feature["name"], value_type=feature["value_type"]
+                    )
+                )
+        # check if features can be found within the feature vector
+        elif model_obj.feature_vector:
+            fv = mlrun.feature_store.common.get_feature_vector_by_uri(
+                model_obj.feature_vector, project=model_endpoint.metadata.project
+            )
+            for feature in fv.status.features.to_dict():
+                feature_set.add_feature(
```
Comment (yaronha) [inline mlrun/api/crud/model_endpoints.py:183]: why `to_dict()` vs `items()` ?

File: mlrun/model_monitoring/model_monitoring_batch.py [hunk 5]
```diff
@@ -479,14 +462,51 @@ def run(self):
                     logger.info(f"{endpoint_id} is router skipping")
                     continue
 
-                df = pd.read_parquet(full_path)
+                # convert feature set into dataframe and get the latest dataset
+                serving_function_name = endpoint.spec.function_uri.replace(
```
Comment (yaronha) [inline mlrun/model_monitoring/model_monitoring_batch.py:466]: use `parse_versioned_object_uri()` instead

**Review by yaronha — APPROVED** [review]

File: mlrun/model_monitoring/helpers.py [hunk 6]
```diff
@@ -56,6 +51,9 @@ def get_model_monitoring_stream_processing_function(
         image="mlrun/mlrun",
     )
 
+    # create monitoring serving graph
+    stream_processor.create_serving_graph(function)
+
```
Comment (Tankilevitch) [inline mlrun/model_monitoring/helpers.py:56]: You are creating serving graph inside a get function, why?

**Review by Tankilevitch — CHANGES_REQUESTED** [review]

File: mlrun/model_monitoring/helpers.py (hunk 6 above)
Comment (Tankilevitch) [reply mlrun/model_monitoring/helpers.py:56]: Lets change it to apply serving graph

File: mlrun/api/crud/model_endpoints.py [hunk 7]
```diff
@@ -59,20 +67,29 @@ def create_or_patch(
...
 
+            # get algorithm from model object if not found in model endpoint object
             if not model_endpoint.spec.algorithm and hasattr(model_obj, "algorithm"):
                 model_endpoint.spec.algorithm = model_obj.algorithm
 
+            # Create monitoring feature set if monitoring found in model endpoint object
+            if model_endpoint.spec.monitoring_mode == "enabled":
+                self.create_monitoring_feature_set(
+                    model_endpoint, model_obj, db_session
+                )
```
Comment (Tankilevitch) [inline mlrun/api/crud/model_endpoints.py:91]: Lets move it to struct that contains the different modes, so we could refer to the available modes.

File: mlrun/api/schemas/model_endpoints.py [hunk 8]
```diff
@@ -32,6 +32,7 @@ class ModelEndpointSpec(ObjectSpec):
     algorithm: Optional[str]
     monitor_configuration: Optional[dict]
     active: Optional[bool]
+    monitoring_mode: Optional[str] = "disabled"
```
Comment (Tankilevitch) [inline mlrun/api/schemas/model_endpoints.py:35]: Change from str to new Model contains the different monitoring modes.

**Review by Tankilevitch — CHANGES_REQUESTED** [review]

File: mlrun/api/crud/model_endpoints.py [hunk 9]
```diff
@@ -691,20 +811,53 @@ def deploy_model_monitoring_stream_processing(
...
             auth_info=auth_info,
         )
 
+        # get the function uri
         function_uri = fn.save(versioned=True)
         function_uri = function_uri.replace("db://", "")
 
+        logger.info(
+            "Deploying model monitoring batch processing function", project=project
+        )
```
Comment (Tankilevitch) [inline mlrun/api/crud/model_endpoints.py:860]: move it near the `_submit_run`

File: mlrun/model_monitoring/helpers.py [hunk 10]
```diff
@@ -41,11 +40,7 @@ def get_model_monitoring_stream_processing_function(
         parquet_batching_max_events=mlrun.mlconf.model_endpoint_monitoring.parquet_batching_max_events,
     )
```
Comment (Tankilevitch) [inline mlrun/model_monitoring/helpers.py:42]: Will have to change the naming of the function

**Pull request merged**